### ISSUE TITLE
ROCm 5.2, OpenCL: wrong codegen with 24bit mul

### ISSUE THREAD

**Issue by preda** [RadeonOpenCompute/ROCm#1764]
Using ROCm 5.2, RadeonVII, Ubuntu 21.10, Linux kernel 5.18.8.
For this OpenCL kernel:
```
kernel void testKernel(global uint *io) {
  uint me = get_local_id(0);
  ulong x = io[me] & 0x00ffffffu;
  io[me] = (x * x) >> 32;
}
```
this ISA is generated:
```
	s_load_dwordx2 s[0:1], s[4:5], 0x0
	v_lshlrev_b32_e32 v0, 2, v0
	v_mov_b32_e32 v2, 0
	s_waitcnt lgkmcnt(0)
	v_mov_b32_e32 v1, s1
	v_add_co_u32_e32 v0, vcc, s0, v0
	v_addc_co_u32_e32 v1, vcc, 0, v1, vcc
	global_store_dword v[0:1], v2, off
	s_endpgm
```
As you see, the generated code disregards the input value, and simply writes 0 (v2). This is not correct, the expected behavior would be to write the high dword of the multiplication, as would be produced by e.g. v_mul_hi_u32_u24

**Comment by dmikushin** [comment]
Hi @preda , thank you for reporting this. I guess you are dumping the assembly using `AMD_OCL_BUILD_OPTIONS_APPEND=-save-temps-all`. Would you be able also to dump LLVM IR, in order to see whether the problem already exists in the IR? Dumping of IR might not be available right away. If so, could you please follow [this instruction](https://github.com/RadeonOpenCompute/ROCm-OpenCL-Runtime/issues/38#issuecomment-342954243)?

**Comment by preda** [comment]
Hi, the value shifted by 32 is an ulong (unsigned 64-bit), the type of "x".

**Comment by preda** [comment]
```
*** IR Dump After Expand reduction intrinsics ***
; Function Attrs: nofree norecurse nounwind
define protected amdgpu_kernel void @testKernel(i32 addrspace(1)* nocapture %0) local_unnamed_addr #8 !kernel_arg_addr_space !155 !kernel_arg_access_qual !156 !kernel_arg_type !157 !kernel_arg_base_type !157 !kernel_arg_type_qual !158 !kernel_arg_name !159 {
  %2 = tail call i32 @llvm.amdgcn.workitem.id.x() #14, !range !35
  %3 = zext i32 %2 to i64
  %4 = getelementptr inbounds i32, i32 addrspace(1)* %0, i64 %3
  %5 = load i32, i32 addrspace(1)* %4, align 4, !tbaa !72
  %6 = and i32 %5, 16777215
  %7 = zext i32 %6 to i64
  %8 = trunc i64 %7 to i32
  %9 = trunc i64 %7 to i32
  %10 = call i32 @llvm.amdgcn.mul.u24(i32 %8, i32 %9)
  %11 = zext i32 %10 to i64
  %12 = lshr i64 %11, 32
  %13 = trunc i64 %12 to i32
  store i32 %13, i32 addrspace(1)* %4, align 4, !tbaa !72
  ret void
}
*** IR Dump After Global Value Numbering ***
; Function Attrs: nofree norecurse nounwind
define protected amdgpu_kernel void @testKernel(i32 addrspace(1)* nocapture %0) local_unnamed_addr #8 !kernel_arg_addr_space !155 !kernel_arg_access_qual !156 !kernel_arg_type !157 !kernel_arg_base_type !157 !kernel_arg_type_qual !158 !kernel_arg_name !159 {
  %2 = tail call i32 @llvm.amdgcn.workitem.id.x() #14, !range !35
  %3 = zext i32 %2 to i64
  %4 = getelementptr inbounds i32, i32 addrspace(1)* %0, i64 %3
  %5 = load i32, i32 addrspace(1)* %4, align 4, !tbaa !72
  %6 = and i32 %5, 16777215
  %7 = zext i32 %6 to i64
  %8 = call i32 @llvm.amdgcn.mul.u24(i32 %6, i32 %6)
  %9 = zext i32 %8 to i64
  store i32 0, i32 addrspace(1)* %4, align 4, !tbaa !72
  ret void
}
```

**Comment by preda** [comment]
It seems the problem occurs during "Global Value Numbering", where the zero appears.

**Comment by dmikushin** [comment]
Interesting. So, before doing `(x * x)`, each `x` is converted to 32-bit:

```
  %8 = trunc i64 %7 to i32
  %9 = trunc i64 %7 to i32
  %10 = call i32 @llvm.amdgcn.mul.u24(i32 %8, i32 %9)
```

**Comment by preda** [comment]
> Interesting. So, before doing `(x * x)`, each `x` is converted to 32-bit:
> 
> ```
>   %8 = trunc i64 %7 to i32
>   %9 = trunc i64 %7 to i32
>   %10 = call i32 @llvm.amdgcn.mul.u24(i32 %8, i32 %9)
> ```
> 
> That's the most exact reason why this code eventually got optimized out as meaningless.

The fact that the arguments to mul.24 are converted to 32bits is normal (because, indeed, mul24 only uses 24bits).

**Comment by dmikushin** [comment]
Right, then (x * x) result is limited by 32 bits. Even if the result is expanded to 64 bits, it could still have only up to 32 meaningful bits, and shifting all of them out becomes zero.

**Comment by preda** [comment]
> Right, then (x * x) result is limited by 32 bits. Even if the result is expanded to 64 bits, it could still have only up to 32 meaningful bits, and shifting all of them out becomes zero.

Yes. So the problem was earlier, when @llvm.amdgcn.mul.u24 was generated, because this is not the right operation as it only produces the low-32 bits of the product. Multiplying two 24-bit values does produce a 48bits result (32+16), and the high-16-bits of those 48 are not captured.

**Comment by preda** [comment]
```
*** IR Dump After Early CSE ***
; Function Attrs: nofree norecurse nounwind
define protected amdgpu_kernel void @testKernel(i32 addrspace(1)* nocapture %0) local_unnamed_addr #8 !kernel_arg_addr_space !155 !kernel_arg_access_qual !156 !kernel_arg_type !157 !kernel_arg_base_type !157 !kernel_arg_type_qual !158 !kernel_arg_name !159 {
  %2 = tail call i32 @llvm.amdgcn.workitem.id.x() #14, !range !35
  %3 = zext i32 %2 to i64
  %4 = getelementptr inbounds i32, i32 addrspace(1)* %0, i64 %3
  %5 = load i32, i32 addrspace(1)* %4, align 4, !tbaa !72
  %6 = and i32 %5, 16777215
  %7 = zext i32 %6 to i64
  %8 = mul nuw nsw i64 %7, %7
  %9 = lshr i64 %8, 32
  %10 = trunc i64 %9 to i32
  store i32 %10, i32 addrspace(1)* %4, align 4, !tbaa !72
  ret void
}
*** IR Dump After AMDGPU IR optimizations ***
; Function Attrs: nofree norecurse nounwind
define protected amdgpu_kernel void @testKernel(i32 addrspace(1)* nocapture %0) local_unnamed_addr #8 !kernel_arg_addr_space !155 !kernel_arg_access_qual !156 !kernel_arg_type !157 !kernel_arg_base_type !157 !kernel_arg_type_qual !158 !kernel_arg_name !159 {
  %2 = tail call i32 @llvm.amdgcn.workitem.id.x() #14, !range !35
  %3 = zext i32 %2 to i64
  %4 = getelementptr inbounds i32, i32 addrspace(1)* %0, i64 %3
  %5 = load i32, i32 addrspace(1)* %4, align 4, !tbaa !72
  %6 = and i32 %5, 16777215
  %7 = zext i32 %6 to i64
  %8 = trunc i64 %7 to i32
  %9 = trunc i64 %7 to i32
  %10 = call i32 @llvm.amdgcn.mul.u24(i32 %8, i32 %9)
  %11 = zext i32 %10 to i64
  %12 = lshr i64 %11, 32
  %13 = trunc i64 %12 to i32
  store i32 %13, i32 addrspace(1)* %4, align 4, !tbaa !72
  ret void
}
```
Now the problem seems to be at "AMDGPU IR optimizations", as before we have a 64-bit MUL of two 24-bit values, which is turned into a 32-bit MUL of 2x 24-bits, which is not correct as it loses the bits above 32.

**Comment by dmikushin** [comment]
Also we could note that [mul24 intrinsic emission](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp#L558) is performed on a type read from the instruction: 

```
bool AMDGPUCodeGenPrepare::replaceMulWithMul24(BinaryOperator &I) const {
...
  Type *Ty = I.getType();
  unsigned Size = Ty->getScalarSizeInBits();
...
Value *Result =
        getMul24(Builder, LHS, RHS, Size, LHSBits + RHSBits, IsSigned);
```

It means at the time when mul is replaced with mul24, the return type has been already changed from 64-bit to 32-bit, which suggests that the cause is not the mul24 itself.

**Comment by b-sumner** [comment]
This doesn't seem to be happening with a main build.  And it looks like https://reviews.llvm.org/D111523 was trying to fix this.  I would have thought ROCm 5.2 would have picked up a commit from last October.

**Comment by preda** [comment]
My report is invalid, I was unawarelly picking up on an old ROCm 3.3.0 that I had lying around. Once I fixed that, and actually used ROCm 5.2.0, the generated ISA became
```
	s_load_dwordx2 s[0:1], s[4:5], 0x0
	v_lshlrev_b32_e32 v0, 2, v0
	s_waitcnt lgkmcnt(0)
	global_load_dword v1, v0, s[0:1]
	s_waitcnt vmcnt(0)
	v_mul_hi_u32_u24_e32 v1, v1, v1
	global_store_dword v0, v1, s[0:1]
	s_endpgm
```
which is what I'd expect. Sorry for that.
